lib/internal/cluster/primary.js

```diff
@@ -158,6 +161,10 @@ function removeHandlesForWorker(worker) {
   });
 }
 
+cluster.getSettings = function() {
```

Comment:
This needs docs

Reply:
@bnoordhuis asked for `getSettings()` to be internal in https://github.com/nodejs/node/issues/49240#issuecomment-1685241747, I understand it should not be documented then? From my side I think it makes sense to make it public and document it, but prefer to wait for his opinion.

Reply:
Yeah, internal stuff should not be in the docs. If there is anything not self-evident about it, please leave code comments, but that goes for everything.

Reply:
The current implementation is public property. If that method is expected to be internal, you need to make it non enumerable or whatever node.js uses to distinguish between public and internal APIs. But I still do not understand why folding a property into an existing object literal property need a new getter to be added. The two are orthogonal. And if you make an PR for the folding and one for the API change, the former will not be encumbered by API changes discussion that always takes time.

Reply:
I have documented the property `getSettings()` so it can be made public. It needs the new getter because the old way to expose the whole object `settings` does not play well with ESM, which makes the `settings` object immutable. I believe the old `settings` should therefore be made deprecated, but have not done so because I proposed it and nobody seconded the option.

Reply:
I'm still having hard time to understand the need to have two properties attached to `cluster` with the very same semantic. If the original implementation breaks with ESM, it then should be fixed to work with ESM and CJS without duplicating it somehow and making the API more messy: 

- cluster.settings -> cluster._settings or cluster.#settings (or whatever node uses to flag a property as internal)
- cluster.settings -> the proposed implementation in the PR 

What is exactly hindering it? And why not tackling the root cause? 
Changing APIs to support ESM does not seem to be a wise idea, nor a global ESM support policy. 

Since the PR approved, I guess it will go in. But I really find APIs changing because of implementation not supported by ESM not a proper solution.

Reply:
With the PR I am trying to solve a practical problem. What follows is my opinionated take. The crux of the matter is that exporting an attribute like `cluster.settings` is a no-no in ESM, and in fact is super ugly in CJS also: you never know if people want to query or set sub-attributes. In ESM it's even worse as exported attributes will be read-only and will not work as expected. I don't see a problem in evolving APIs to work in ESM; usually they were not good choices for CJS either.

The proper way is to have definite getters and setters, so in this case we should deprecate `settings` and leave `getSettings()`. But the current form of the PR solves the issue at hand with minimal changes.

Reply:
I think having ESM and CJS support without changing API is better when doable. Changing public API in general without strong rationale should be avoided in general. I would not put ESM support a rationale strong enough to change a public API.

`getSettings()` is introduced, then `settings` is deprecated.

And here can be achieved by doing `settings` a getter: `get settings () {}` (or using other syntax). It's a breaking change if used to change the values. But nobody is supposed to use it that way. The code is not even able to handle dynamic change of cluster.settings at runtime once workers are created. So a breaking change that actually enforce the API usage the right way, as you also stated.  

Since it's not a widely used API, nor , it's not a big deal. And I can live as an API consumer in a lib I maintain with a `try {} catch {}` around `getSettings()` to ensure ESM/CJS support on several node.js version. But these comments maybe can be used to initiate a reflexion on ESM support policy in node.js to avoid API consumer to use ugly hack.

Reply:
Hmmm, what induces you to think that this PR is an ugly hack? It is indeed a pretty elegant solution if I may so, only deprecating the public use of `settings` is in fact missing.

Reply:
> Hmmm, what induces you to think that this PR is an ugly hack? It is indeed a pretty elegant solution if I may so myself, only deprecating the public use of `settings` is in fact missing to be 100% right at least for me.

A real elegant solution would have not changed a public API to support ESM ;)

And I've not written that PR is an ugly hack ;) I've written that supporting ESM/CJS in code using cluster API (the API consumer) and expected to run on various node.js versions will require ugly hacks. And it could be avoided. 

Anyhow, ESM support in cluster is still buggy. Worker as ESM module fails to init.

Reply:
> A real elegant solution would have not changed a public API to support ESM ;)

Good luck with that.

> And I've not written that PR is an ugly hack ;) I've written that supporting ESM/CJS in code using cluster API (the API consumer) and expected to run on various node.js versions will require ugly hacks. And it could be avoided.

I await your proposal.

> Anyhow, ESM support in cluster is still buggy. Worker as ESM module fails to init.

It works fine for me :thinking:

Reply:
> > And I've not written that PR is an ugly hack ;) I've written that supporting ESM/CJS in code using cluster API (the API consumer) and expected to run on various node.js versions will require ugly hacks. And it could be avoided.
> 
> I await your proposal.
> 

`cluster.prototype.settings = function() {}`/`Object.defineProperty(settings, { get: function () {}})` should works with ESM and CJS. 
Is it prohibited in node.js internal JS code? 

> > Anyhow, ESM support in cluster is still buggy. Worker as ESM module fails to init.
> 
> It works fine for me 🤔

Not if you use a dedicated ESM file as a cluster worker: https://github.com/nodejs/node/issues/48578

Reply:
> `cluster.prototype.settings = function() {}`/`Object.defineProperty(cluster, 'settings', { get: function () {}})` should works with ESM and CJS. Is it prohibited in node.js internal JS code?

You are aware that this change breaks backwards compatibility in CJS as it does not allow modifying `cluster.settings`, as it used to do.

> > > Anyhow, ESM support in cluster is still buggy. Worker as ESM module fails to init.
> > 
> > 
> > It works fine for me 🤔
> 
> Not if you use a dedicated ESM file as a cluster worker: #48578

Oh, of course my PR doesn't fix all issues. It fixes setting `schedulingPolicy` as stated.